### PR TITLE
fix: fixed the issue where the traffic light button on the white thumbnail is difficult to see in dark mode.

### PR DESCRIPTION
**Pull request by Allsochen**
Fixed the issue where the traffic light button on the white thumbnail is difficult to see in dark mode.

Fixed the issues https://github.com/lwouis/alt-tab-macos/issues/2892

In dark mode, there are two situations where it is difficult to distinguish:
- The thumbnail is gray, and the traffic light button appears faded.
- The thumbnail is white, and the traffic light button is not visible at all.

image

image

After fixed:
- Fixed the transparent icon
image
image

- Fix the fullscreen/defullsreen icon
image
image

### CODE REVIEW

**Comment by Allsochen** [comment]
> It seems it could be done more directly by forcing the appearance of `TrafficLightButton` to aqua.

Thank you for the reminder. I re-implemented it on TrafficLightButton, but encountered an issue: adding a background to TrafficLightButton didn't work. Ultimately, changing the inheritance of TrafficLightButton to NSView solved the problem.

**Comment by Allsochen** [comment]
> I'm not sure I follow you, but I was thinking of something like this:
> 
> ```swift
> class TrafficLightButton: NSButton {
>     [...]
>     init(_ type: TrafficLightButtonType, _ tooltip: String, _ size: CGFloat) {
>         [...]
>         appearance = .init(named: .aqua)
>     }
>     [...]
> }
> ```

I don't know how to fix it using `aqua`, but I fixed it another way by extending NSView.

File: src/logic/Preferences.swift [hunk 1]
```diff
@@ -104,8 +104,8 @@ class Preferences {
     // not exposed as preferences now but may be in the future, probably through macro preferences
     static var fontColor: NSColor { .white }
     static var windowPadding: CGFloat { 18 }
-    static var interCellPadding: CGFloat { 5 }
-    static var intraCellPadding: CGFloat { 5 }
+    static var interCellPadding: CGFloat { 15 }
+    static var intraCellPadding: CGFloat { 8 }
```
Comment (lwouis) [inline src/logic/Preferences.swift:108]: I'm not sure to understand this change. Is it related to the goal of the MR?

**Comment by Allsochen** [comment]
> Hi everyone,
> 
> Indeed, I think this MR is quite large, for its intended goal. I was able to reproduce the issue by switching to Dark Mode. Then I applied [the fix suggested by @decodism](https://github.com/lwouis/alt-tab-macos/pull/3433#issuecomment-2171770891), and it fixes the issue for me. One line of code seems to fix the issue:
> 
> image

This MR not only fixes the issue with the transparency of the traffic lights but also does the following:

- Fixes the shape of fullscreen/defullscreen buttons.
- Adds code comments to make it easier to understand.

File: src/logic/Preferences.swift (hunk 1 above)
Comment (Allsochen) [reply src/logic/Preferences.swift:108]: This fix is unrelated to the changes here. The modification was made because many AltTab users around me have reported that the spacing between thumbnails is too small, so I changed it to the thumbnail spacing of Windows 11. If you disagree, we can leave this part unchanged for now and consider revising it later as a separate advanced configuration option.

**Comment by Allsochen** [comment]
> Hi everyone,
> 
> Indeed, I think this MR is quite large, for its intended goal. I was able to reproduce the issue by switching to Dark Mode. Then I applied [the fix suggested by @decodism](https://github.com/lwouis/alt-tab-macos/pull/3433#issuecomment-2171770891), and it fixes the issue for me. One line of code seems to fix the issue:
> 

This MR not only fixes the issue with the transparency of the traffic lights button but also does the following:

Fixes the shape of fullscreen/defullscreen buttons.
Adds code comments to make it easier to understand.

File: src/logic/Preferences.swift (hunk 1 above)
Comment (Allsochen) [reply src/logic/Preferences.swift:108]: 83ab94fdb3b446c1e98e5f63bc2f60ef

Changed the thumbnail spacing equals to Windows 11 spacing.

File: src/ui/generic-components/TrafficLightButton.swift [hunk 2]
```diff
@@ -1,219 +1,379 @@
...
-    }
-
+    /// Mouse entered event handler.
     override func mouseEntered(with event: NSEvent) {
         isMouseOver = true
-        setNeedsDisplay()
+        setNeedsDisplay(bounds)
     }
 
+    /// Mouse exited event handler.
```
Comment (lwouis) [inline src/ui/generic-components/TrafficLightButton.swift:50]: I don't think these comments are the way to go. When we see a method whith `override`, we can use our IDE to display the original method's documentation, very easily:

image

File: src/ui/generic-components/TrafficLightButton.swift [hunk 3]
```diff
@@ -1,219 +1,379 @@
...
+
+        let disk = NSBezierPath()
+        disk.appendOval(in: NSMakeRect(bounds.origin.x + 0.5, bounds.origin.y + 0.5, bounds.width - 1, bounds.height - 1))
+
+        // Fill the disk with the black background color
+        NSColor.black.setFill()
+        disk.fill()
+
+        // Fill the disk with the background color
+        diskBackgroundColor.draw(in: disk, relativeCenterPosition: .zero)
```
Comment (lwouis) [inline src/ui/generic-components/TrafficLightButton.swift:97]: It seems you added drawing a black disk here. What issue is it fixing?

File: src/ui/generic-components/TrafficLightButton.swift [hunk 4]
```diff
@@ -1,219 +1,379 @@
...
+    /// - Parameters:
+    ///   - disk: The `NSBezierPath` representing the button's disk.
     private func drawDimming(_ disk: NSBezierPath) {
+        // Set the line width of the disk
         disk.lineWidth = 1
-        if (isHighlighted) {
-            NSColor.black.withAlphaComponent(0.5).setFill()
-            disk.fill()
-        } else if (isMouseOver) {
+        if isMouseOver {
```
Comment (lwouis) [inline src/ui/generic-components/TrafficLightButton.swift:113]: You removed the difference of being highlighted = 50% opacity, and being hovered by the mouse = 25% opacity. Now the mouse hover is 25%, and the highlight is gone; it's always 100%. I'm not sure why this change is made. How is it related to the stated goal of the MR?

File: src/ui/generic-components/TrafficLightButton.swift [hunk 5]
```diff
@@ -1,219 +1,379 @@
...
-            window_?.toggleFullscreen()
-        } else if (type == .miniaturize) {
-            window_?.minDemin()
-        } else if (type == .close) {
-            window_?.close()
-        } else if (type == .quit) {
-            window_?.application.quit()
-        }
-    }
-
```
Comment (lwouis) [inline src/ui/generic-components/TrafficLightButton.swift:34]: Was this really not necessary? I'm thinking that if it's there, there may be a reason. Have you tested thoroughly? Also checking the git history to see when and why this was added?

File: src/ui/generic-components/TrafficLightButton.swift [hunk 6]
```diff
@@ -1,219 +1,379 @@
...
+    /// Mouse entered event handler.
     override func mouseEntered(with event: NSEvent) {
         isMouseOver = true
-        setNeedsDisplay()
+        setNeedsDisplay(bounds)
```
Comment (lwouis) [inline src/ui/generic-components/TrafficLightButton.swift:47]: What is the purpose of this change?

File: src/ui/main-window/ThumbnailView.swift [hunk 7]
```diff
@@ -95,6 +95,8 @@ class ThumbnailView: NSStackView {
                         yOffset += ThumbnailView.windowsControlSize + ThumbnailView.windowsControlSpacing
                     }
                 }
+                // Force the icon to repaint
+                icon.display()
```
Comment (lwouis) [inline src/ui/main-window/ThumbnailView.swift:99]: Why is this necessary? What was failing before we added this new call?

File: src/logic/Preferences.swift (hunk 1 above)
Comment (lwouis) [reply src/logic/Preferences.swift]: I think changes shouldn't be bundled. It can create confusion while reviewing MRs. If you propose to make multiple unrelated changes, I suggest opening multiple MRs, each with their goal. They can be reviewed with more clarity.

Furthermore, we use a convention on commits. You can learn [a bit more here](https://alt-tab-macos.netlify.app/contributing#raising-a-pull-request). One commit = 1 change. Here there should be a `fix:` for the traffic light opacity issue, since it's indeed a bug fix. Then there should be a `feat:` to update the spacings, since it's not a bug fix bug an improvement. Finally, for documentation, one could use `docs:` or `refactor:` for the commit which documents the existing code.

Regarding the change of spacing: I think Windows 11 Spacings should live in a Windows 11 theme. We already support Themes in Preferences > Appearance. Let's add a Windows 11 theme. Other people were asking about it I believe.

File: src/ui/main-window/ThumbnailView.swift (hunk 7 above)
Comment (Allsochen) [reply src/ui/main-window/ThumbnailView.swift:99]: If this modification is not added, an issue arises: after the user clik the fullscrren button to enters full screen, the fullscreen button does not redraw, resulting in the user still seeing the button as if it were in fullscreen mode, rather than the defullscreen button. To make this issue clearer to you, I will open a new MR, so you can see the problem and the solution more clearly.

File: src/ui/generic-components/TrafficLightButton.swift (hunk 6 above)
Comment (Allsochen) [reply src/ui/generic-components/TrafficLightButton.swift:47]: In terms of the effect, their performance is the same. This class inherits from the NSView class, and through this modification, we can specify a fixed area for redrawing, which will be more precise.

File: src/ui/generic-components/TrafficLightButton.swift (hunk 4 above)
Comment (Allsochen) [reply src/ui/generic-components/TrafficLightButton.swift:113]: The button is currently an NSView, and when it needs to be displayed, it should show its color at 100% opacity. When the mouse hovers over it, it should display at 25% opacity. The NSButton has an isHighlighted state, but this state is no longer needed with the current setup.

File: src/ui/generic-components/TrafficLightButton.swift (hunk 3 above)
Comment (Allsochen) [reply src/ui/generic-components/TrafficLightButton.swift:97]: This is the key to fixing the issue of transparent button in the dark theme. The black disk serves as a background to prevent the buttons from being transparent.

File: src/ui/generic-components/TrafficLightButton.swift (hunk 5 above)
Comment (Allsochen) [reply src/ui/generic-components/TrafficLightButton.swift:34]: Thank you for the reminder, this can be changed to mouseDown.

**Comment by lwouis** [comment]
I'll grab the change on the `dev` branch. It will be in the next release 👍
